pkg/agent/core/ngt/handler/grpc/index_test.go

```diff
@@ -508,106 +508,538 @@ func Test_server_CreateIndex(t *testing.T) {
+				args: args{
+					in: emptyPayload,
+				},
+				fields: fields{
+					srvOpts:   defaultSrvOpts,
+					svcCfg:    defaultSvcCfg,
+					svcOpts:   defaultSvcOpts,
+					indexPath: mkdirTemp(),
+				},
+				beforeFunc: func(t *testing.T, ctx context.Context, s Server, n service.NGT) {
+					if _, err := s.MultiInsert(ctx, irs); err != nil {
+						t.Error(err)
+					}
+					// we need to create index before saving to store the indexed vector
+					if _, err := s.CreateIndex(ctx, &payload.Control_CreateIndexRequest{
+						PoolSize: 1,
+					}); err != nil {
+						t.Error(err)
+					}
+				},
+				want: want{
+					wantRes: emptyPayload,
+				},
+				checkFunc: func(test test, ctx context.Context, s Server, n service.NGT, w want, e *payload.Empty, err error) error {
+					if err := defaultCheckFunc(test, ctx, s, n, w, e, err); err != nil {
+						return err
+					}
+					if err := checkBackupFolder(test.fields, ctx, irs.Requests); err != nil {
+						return err
+					}
+
+					return nil
+				},
+			}
+		}(),
+		func() test {
+			// bulk insert request to make saveIndex cost time
+			insertNum := 100000
+			irs, err := request.GenMultiInsertReq(request.Float, vector.Gaussian, insertNum, dim, nil)
+			if err != nil {
+				t.Error(err)
+			}
+
+			return test{
+				name: "Equivalence Class Testing case 3.1: success to save index when other save index process is running",
+				args: args{
+					in: emptyPayload,
+				},
+				fields: fields{
+					srvOpts:   defaultSrvOpts,
+					svcCfg:    defaultSvcCfg,
+					svcOpts:   defaultSvcOpts,
+					indexPath: mkdirTemp(),
+				},
+				beforeFunc: func(t *testing.T, ctx context.Context, s Server, n service.NGT) {
+					if _, err := s.MultiInsert(ctx, irs); err != nil {
+						t.Error(err)
+					}
+					// we need to create index before saving to store the indexed vector
+					if _, err := s.CreateIndex(ctx, &payload.Control_CreateIndexRequest{
+						PoolSize: uint32(insertNum),
+					}); err != nil {
+						t.Error(err)
+					}
+
+					// execute save index twice
+					go s.SaveIndex(ctx, emptyPayload)
+				},
+				checkFunc: func(t test, ctx context.Context, s Server, n service.NGT, w want, e *payload.Empty, err error) error {
+					if err := defaultCheckFunc(t, ctx, s, n, w, e, err); err != nil {
+						return err
+					}
+
+					timeout := time.After(10 * time.Second)
+					ticker := time.Tick(10 * time.Millisecond)
+					for {
+						select {
+						case <-timeout:
+							// execute save index twice
+							go s.SaveIndex(ctx, emptyPayload)
+
+							gotRes, err := s.SaveIndex(ctx, emptyPayload)
+							if err := defaultCheckFunc(t, ctx, s, n, w, gotRes, err); err != nil {
+								return err
+							}
+						case <-ticker:
+							// execute save index twice
+							go s.SaveIndex(ctx, emptyPayload)
+
+							gotRes, err := s.SaveIndex(ctx, emptyPayload)
+							if err := defaultCheckFunc(t, ctx, s, n, w, gotRes, err); err == nil {
+								return nil
+							}
+						}
+					}
```

Comment:
Can you please explain the background of this implementation? 🙏

Reply:
since we cannot guarantee the first saveIndex call on line 773 is executed asynchronously when we execute the actual saveIndex call on line 1041, I wanted to check it multiple times when it is not happen.